### PR TITLE
Unit testing

### PR DESCRIPTION
**Pull request by tearzday**
1. Task: [link](https://github.com/rolling-scopes-school/tasks/blob/master/react/modules/tasks/tests.md)
2. Screenshot:
3. Deploy: image
4. Done 17.07.2025 / deadline 24.07.2025
5. Score: 55 / 70
  - [x] (15/15) Test Runner Setup (Jest/Vitest configured with RTL, coverage reporting works)
  - [x] (20/20) Search Component Tests (localStorage, user input, search functionality)
  - [x] (15/15) Results/CardList Component Tests (rendering, data display, loading states)
  - [x] (15/15) Card Component Tests (rendering)
  - [x] (15/15) Error Boundary Tests (error catching, fallback UI, test error button)
  - [x] (15/15) API Integration Tests (mocked API calls, success/error handling)
  - [x] (5/5) Husky Pre-push Hook (tests run automatically before push)

### CODE REVIEW

**Comment by Margaryta-Maletz** [comment]
image
Why is there 55/70? Must be 100/100!

File: src/api/api.test.ts [hunk 1]
```diff
@@ -0,0 +1,40 @@
+import { dataCards } from '@/__tests__/__mocks__/MockCard';
+import APICard from './card';
+
+describe('Test API', () => {
```
Comment (Margaryta-Maletz) [inline src/api/api.test.ts:4]: Use beforeEach to reset mocks before each test to avoid test interference.
```
beforeEach(() => {
  vi.resetAllMocks();
});
```

**Review by Margaryta-Maletz — APPROVED** [review]
Great job!
But some comments on how to improve your code you might want to consider and fix)
Score: 100 / 100
  - [x] (15/15) Test Runner Setup (Jest/Vitest configured with RTL, coverage reporting works)
  
  - [x] (20/20) Search Component Tests (localStorage, user input, search functionality)
  Good job, but you can improve these test cases:
    - Add tests to ensure localStorage is cleared or updated correctly when the input changes or the search is performed multiple times.
    - Add tests for edge cases, such as invalid input (e.g., special characters or excessively long strings).

    One example I added, you can do the another yourself.
    
  - [x] (15/15) Results/CardList Component Tests (rendering, data display, loading states)
    
  - [x] (15/15) Card Component Tests (rendering)
    List of improvements:
    - Add tests for edge cases, such as missing or null props, and ensure the component handles them gracefully.
    - Add tests to verify accessibility attributes, such as alt text for images and ARIA roles.
    - Include snapshot tests to ensure the rendered output does not change unexpectedly.
    There is only recommendation. You can easily skip it.
    
  - [x] (15/15) Error Boundary Tests (error catching, fallback UI, test error button)
    You can add tests to verify the fallback UI's content, such as the error message and refresh button functionality.
    
  - [x] (15/15) API Integration Tests (mocked API calls, success/error handling)
     List of improvements:
     - Add tests for edge cases, such as empty responses or unexpected data formats.
     - Add tests for network errors or timeouts to ensure robust error handling.
     As usually, there is only recommendation. You can easily skip it.
     
  - [x] (5/5) Husky Pre-push Hook (tests run automatically before push)

File: src/components/search/Search.test.tsx [hunk 2]
```diff
@@ -0,0 +1,135 @@
+import Search from './Search';
+import { render, screen, waitFor } from '@testing-library/react';
+import { userEvent } from '@testing-library/user-event';
+
+describe('Test Search Сomponent', () => {
```
[the remark this answers is not in the document]
Comment (tearzday) [reply src/components/search/Search.test.tsx:5]: Aww, thanks for noticing!
